### ISSUE TITLE
Saving the RLMSyncUser Instance for offline login

### ISSUE THREAD

**Issue by Montrazul** [realm/realm-mobile-platform#76]
My App needs to support offline login (only if the user at least logged in once before with a internet connection on the current device). For that I wanted to cache the `RLMSyncUser` object which is returned by the authentication service (online authentication). 

Unfortunately `RLMSyncUser` is not implementing `NSCoding` so that I cant save the `RLMSyncUser` in the `UserDefaults`. 

Whats the recommended way to cache `RLMSyncUser` objects for offline login?

**Comment by bigfish24** [comment]
The SDK automatically caches the user in the keychain after first login. You can then retrieve it via [`currentUser](https://realm.io/docs/swift/latest/api/Extensions/SyncUser.html#/s:ZvE10RealmSwiftCSo11RLMSyncUser7currentGSqS0__) property or alternatively if your app supports multiple users [`all`](https://realm.io/docs/swift/latest/api/Extensions/SyncUser.html#/s:ZvE10RealmSwiftCSo11RLMSyncUser3allGVs10DictionarySSS0__)

**Comment by Montrazul** [comment]
It supports multiple users. So if i am logged out and im offline, the user inputs his email address and passwort. How can I know get the related user from `SyncUser.all`. And if im logged out: isnt `currentUser` nil?

**Comment by bigfish24** [comment]
If you are logged out then `currentUser` is `nil`. If you support multiple users you should use `currentUser`. Instead `all` returns a dictionary with the key being the user Id (email for your example) and the value the `SyncUser`.

**Comment by Montrazul** [comment]
Sorry, dont get it. 

> If you are logged out `currentUser is nil`. If you support multiple users you should use `currentUser`.

So how can I use `currentUser` if it's (like you just said) nil when im logged out? If I register using the email as username, the email is the key for the `all` dictionary. So when im logged out I just have to check the `all` dictionary if there exists a value for the given email key. So I would not need `currentUser` at all (ofcourse except for the use of the realm object, there ofcourse I need the currentUser).

**Comment by stel** [comment]
`SyncUser.all` keeps only valid logged in users. If you log out the user, the corresponding `SyncUser` will be removed from that dictionary.

There is no way to log in the user without network connection. You can only use previously logged in users (either from `SyncUser.all` or `SyncUser.currentUser` in case of a single user).

**Comment by Montrazul** [comment]
Wow.... this is hard.... i cant login offline.... that message just told me that i cant use the great realm mobiel platform.....

**Comment by bigfish24** [comment]
Sorry for the struggle with communication, but if you call `logout` on a user then it will be removed from `all`, however if the user just exits the app but that never gets call then the user is stored and can be used again to open a synchronized Realm.

We currently don't offer an offline login (note this means starting from nothing, rather than starting from a persisted logged in user in `all` to open a synchronized Realm), but it is on the roadmap. We call this an "anonymous user" since you can create it offline and immediate start using a synchronized realm. Maybe that's what you are looking for?

**Comment by Montrazul** [comment]
Ok. So It's rescued. I wont perform a `logout` on a `SyncUser` and will perform the offline login using another way.

Our App also currently does not support offline registration and for the first login on a device there is also a working connection required. But all other logins should work without connection. So that is the way to go I guess.

If you are done with your "anonymous user" I will be happy to implement offline registration / login for my Users.

**Comment by Montrazul** [comment]
No. Because our App is never allowed to automatically Login after App Restart or something similar. If the App is closed and the User opens it up he always has to authenticate. No Auto-Login.

**Comment by bigfish24** [comment]
>No. Because our App is never allowed to automatically Login after App Restart or something similar. If the App is closed and the User opens it up he always has to authenticate. No Auto-Login. Otherwise it would be possible to have unauthorized access to sensible data.

Now I am confused, why do you want to persist the user if this is the case? To accomplish this you will then have to logout whenever the user exits the app to ensure Realm removes their credentials from the keychain.

**Comment by Montrazul** [comment]
What if you are in an Area where the connection to the internet is not given for 100% when you are on your mobile device. What if you really need to login to your App right now but you have no Internet Connection? But what if you also need to make sure that only the user has access to the app who really is the owner of the account?

So you are offline. The User has to login to get access to the data but he also has to authenticate. So he enters his email address and his user password. I now somehow have to make sure, that email and password are the correct combination. For that I stored email and password (also the identity of the realm) in the `UserDefaults` after the user logged in for the first time with working internet connection. The user is now authenticated and can change data. Since Realm Mobile Platform is offline first its the perfect solution for that scenario. 

So before each login I check if there is a working connection to the internet. If yes I use the realm login. If no, I use the offline login using the `UserDefaults` and also load the related `SyncUser` from `SyncUser.all`.

Now the user is working with the app and is changing data. What if connection comes back while hes working with the app? For that I needed to persist the `SyncUser`. That is where `SyncUser.all` comes into play. My Realm always initializes with the `SyncUser` of the currently logged in User. So when the connection comes back the synchronization can continue working without needing a relog or something else.

Does it make sense what im saying?

**Comment by bigfish24** [comment]
Sorry for slow reply, but I am having trouble following. When you call:

```swift
public static func logIn(with credentials: SyncCredentials,
                             server authServerURL: URL,
                             timeout: TimeInterval = 30,
                             onCompletion completion: @escaping UserCompletionBlock)
```

if the completion block fires and the login was successful, the `SyncUser` is persisted to the keychain. If you are using a single user, you can retrieve this `SyncUser` offline on the next app launch via: `SyncUser.current` and then use that to open the synchronized Realm offline. When the internet connection reestablishes the SDK will automatically synchronize any changes made to the Realm while offline for you.

(If you login multiple users in the app then they are retrievable via `SyncUser.all`)

You can definitely add your own "offline login" by storing the email and password, but I would advise against the `UserDefaults` as this is insecure. Instead you can store this in the keychain, and then validate the "offline login" against those values.

Once you call `aUser.logOut()` the `SyncUser` will be removed from the keychain.

**Comment by srpadronm** [comment]
I tested retrieving the SyncUser offline using SyncUser.current, but it doesn't automatically update the server once the internet connection is back. I need to relaunch the app to be able to synchronise all the updated made offline. 
Please could you advise.

**Comment by bigfish24** [comment]
@srpadronm could you provide more details? What I understand from your comment is that you retrieve the `SyncUser` offline and use that to open the Realm, then apply changes to the Realm, but after you come back online the Realm doesn't start syncing again?

Or are you trying to do something with `SyncUser`?

A code sample of the `SyncUser` and opening Realm logic would be helpful to debug.

**Comment by srpadronm** [comment]
Correct, I retrieve the SyncUser offline and use it to open the Realm, then apply changes, but the Realm doesn't start syncing again after the app goes back online. I need to relaunch the app to be able to synchronise all the updates made offline.

` 
    
    func logIn(completion: ((Void)->Void)? = nil) {
        guard user == nil else {
            completion?()
            return
        }
        
        let credentials = SyncCredentials.usernamePassword(username: username, password: password)
        SyncUser.logIn(with: credentials, server: authServerURL, timeout: 30) {
            user, error in
            if let user = user {
                self.user = user
                DispatchQueue.main.async {
                    completion?()
                }
            
            } else if let user = SyncUser.current {
                self.user = user
                DispatchQueue.main.async {
                    completion?()
                }
            } else if let error = error {
                fatalError("Could not log in: \(error)")
            }
        }
    }
    
    func realm() -> Realm? {
        if let user = user {
            let syncConfig = SyncConfiguration(user: user, realmURL: syncServerURL)
            let config = Realm.Configuration(syncConfiguration: syncConfig)
            guard let realm = try? Realm(configuration: config) else {
                fatalError("Could not load Realm")
            }
            return realm
        } else {
            return nil
        }
    }`

**Comment by bigfish24** [comment]
Yeah the issue here is a timing problem. I am assuming you call `logIn` and then immediately after `realm`, however, `SyncUser.logIn` is asynchronous, so there is a time delay for the login to happen before the callback will fire. This means `self.user = user` occurs after you already called `realm` and attempted to open the Realm, resulting in the opening failing because `user` is `nil`.

You need to wait for the login to complete and ensure you have an actual `SyncUser` object before opening the Realm.

We plan to offer support to create "anonymous users" offline synchronously, but it hasn't been implemented yet.

**Comment by srpadronm** [comment]
@bigfish24 I believe this is not the problem. `realm` is call only after I have an actual `SyncUser` 

If I open the app offline, I can retrieve the `SyncUser.current` and open the Realm, without errors, but all updates I made offline (e.g adding new records) are not automatically synchronised with the server once the network connection is back. I need to close and open the app again to be able to synchronise all the updates made offline.

I created a repository with a small example of my code : https://github.com/srpadronm/RMP_Test

**Comment by teotwaki** [comment]
@stel @jpsim Can I get your eyes on this? Thanks

**Comment by stel** [comment]
Sync session uses incremental backoff in order to reconnect after the network is available, so currently  reconnection doesn't happen immediately. The possible solution for this is to use Apple's Reachability API, some work is done [here](https://github.com/realm/realm-object-store/pull/289) but the api isn't designed yet.

**Comment by stel** [comment]
@srpadronm please try to enable all logs and check if there any errors:

```Swift
SyncManager.shared.logLevel = .all
```

**Comment by srpadronm** [comment]
@stel 
Please see below the logs I get by running the app with internet connection:

> 2017-01-23 18:53:06.967293 RMPTest[2177:456331] Sync: Opening Realm file: /var/mobile/Containers/Data/Application/02C61C4E-8A73-4EB7-83C2-5205C0AE6E7D/Documents/realm-object-server/8e55789bac5b692cb576c72fa6b33f39/realm%3A%2F%2F159.203.106.195%3A9080%2F%7E%2Frealmtasks
2017-01-23 18:53:06.968357 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::Websocket()
2017-01-23 18:53:06.968558 RMPTest[2177:456331] Sync: Connection[1]: Reconnecting in 0 milliseconds
2017-01-23 18:53:06.968753 RMPTest[2177:456331] Sync: Connection[1]: Session[1]: Starting session for '/var/mobile/Containers/Data/Application/02C61C4E-8A73-4EB7-83C2-5205C0AE6E7D/Documents/realm-object-server/8e55789bac5b692cb576c72fa6b33f39/realm%3A%2F%2F159.203.106.195%3A9080%2F%7E%2Frealmtasks'
2017-01-23 18:53:06.968891 RMPTest[2177:456331] Sync: Connection[1]: Session[1]: last_version_available  = 34
2017-01-23 18:53:06.969031 RMPTest[2177:456331] Sync: Connection[1]: Session[1]: progress_server_version = 20
2017-01-23 18:53:06.969232 RMPTest[2177:456331] Sync: Connection[1]: Session[1]: progress_client_version = 32
2017-01-23 18:53:06.969458 RMPTest[2177:456331] Sync: Connection[1]: Resolving '159.203.106.195:9080'
2017-01-23 18:53:06.970659 RMPTest[2177:456331] Sync: Connection[1]: Connecting to endpoint '159.203.106.195:9080' (1/1)
2017-01-23 18:53:07.065657 RMPTest[2177:456331] Sync: Connection[1]: Connected to endpoint '159.203.106.195:9080' (from '192.168.1.82:65070')
2017-01-23 18:53:07.065793 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::initiate_client_handshake()
2017-01-23 18:53:07.066201 RMPTest[2177:456331] Sync: Connection[1]: HTTP request =
GET /realm-object-server HTTP/1.1

> Host: 159.203.106.195

> Upgrade: websocket

> Connection: Upgrade

> Sec-WebSocket-Key: AhlDlPt/ywbAmFT+rus4VQ==

> Sec-WebSocket-Version: 13

> Sec-WebSocket-Protocol: io.realm.protocol
2017-01-23 18:53:07.066703 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::read_http_message_line()
2017-01-23 18:53:07.165233 RMPTest[2177:456331] Sync: Connection[1]: Websocket: http_line = HTTP/1.1 101 Switching Protocols
2017-01-23 18:53:07.165565 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::read_http_message_line()
2017-01-23 18:53:07.166013 RMPTest[2177:456331] Sync: Connection[1]: Websocket: http_line = upgrade: websocket
2017-01-23 18:53:07.166304 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::read_http_message_line()
2017-01-23 18:53:07.166539 RMPTest[2177:456331] Sync: Connection[1]: Websocket: http_line = connection: Upgrade
2017-01-23 18:53:07.166665 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::read_http_message_line()
2017-01-23 18:53:07.167075 RMPTest[2177:456331] Sync: Connection[1]: Websocket: http_line = sec-websocket-accept: InqLnR9wV9INQbLXzwBv2WpM814=
2017-01-23 18:53:07.167651 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::read_http_message_line()
2017-01-23 18:53:07.167890 RMPTest[2177:456331] Sync: Connection[1]: Websocket: http_line = sec-websocket-protocol: io.realm.protocol
2017-01-23 18:53:07.168099 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::read_http_message_line()
2017-01-23 18:53:07.168235 RMPTest[2177:456331] Sync: Connection[1]: Websocket: http_line =
2017-01-23 18:53:07.168568 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::handle_http_response_received()
2017-01-23 18:53:07.169064 RMPTest[2177:456331] Sync: Connection[1]: Sending: CLIENT(protocol_version=15, client_info_size=2)
2017-01-23 18:53:07.169380 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::frame_reader_loop()
2017-01-23 18:53:07.169692 RMPTest[2177:456331] Sync: Connection[1]: stage_init
2017-01-23 18:53:07.170013 RMPTest[2177:456331] Sync: Connection[1]: WebSocket: frame_reader.read_size = 2
2017-01-23 18:53:07.170759 RMPTest[2177:456331] Sync: Connection[1]: Session[1]: Sending: BIND(server_path='/8e55789bac5b692cb576c72fa6b33f39/realmtasks', signed_user_token_size=597, need_file_ident_pair=0)
2017-01-23 18:53:07.171047 RMPTest[2177:456331] Sync: Connection[1]: handle_write_output_buffer
2017-01-23 18:53:07.171485 RMPTest[2177:456331] Sync: Connection[1]: Session[1]: Sending: IDENT(server_file_ident=429741678585699184, client_file_ident=2, client_file_ident_secret=1012524929554286685, scan_server_version=20, scan_client_version=32, latest_server_version=20, latest_server_session_ident=7877148332545424738)
2017-01-23 18:53:07.171820 RMPTest[2177:456331] Sync: Connection[1]: handle_write_output_buffer
2017-01-23 18:53:07.171968 RMPTest[2177:456331] Sync: Using already open Realm file: /var/mobile/Containers/Data/Application/02C61C4E-8A73-4EB7-83C2-5205C0AE6E7D/Documents/realm-object-server/8e55789bac5b692cb576c72fa6b33f39/realm%3A%2F%2F159.203.106.195%3A9080%2F%7E%2Frealmtasks
2017-01-23 18:53:07.172548 RMPTest[2177:456331] Sync: Connection[1]: Session[1]: Sending: UPLOAD(client_version=34, server_version=20, changeset_size=30, timestamp=65127039315)
2017-01-23 18:53:07.173122 RMPTest[2177:456331] Sync: Connection[1]: Session[1]: Changeset: 05 00 03 0D 0E 01 0E 00 06 02 00 0E 01 6E 06 01 01 0E 00 05 00 04 1E 02 00 03 20 0E 0E 0E
2017-01-23 18:53:07.173465 RMPTest[2177:456331] Sync: Connection[1]: handle_write_output_buffer
2017-01-23 18:53:07.405759 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::frame_reader_loop()
2017-01-23 18:53:07.406222 RMPTest[2177:456331] Sync: Connection[1]: stage_header_beginning
2017-01-23 18:53:07.406840 RMPTest[2177:456331] Sync: Connection[1]: WebSocket: frame_reader.read_size = 49
2017-01-23 18:53:07.407391 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::frame_reader_loop()
2017-01-23 18:53:07.407955 RMPTest[2177:456331] Sync: Connection[1]: stage_payload
2017-01-23 18:53:07.408445 RMPTest[2177:456331] Sync: Connection[1]: websocket message received: download 1 21 34 21 7877148332545424738 34 0 0 0
2017-01-23 18:53:07.409001 RMPTest[2177:456331] Sync: Connection[1]: message_type = download
2017-01-23 18:53:07.409898 RMPTest[2177:456331] Sync: Connection[1]: Session[1]: Received: DOWNLOAD(scan_server_version=21, scan_client_version=34, latest_server_version=21, latest_server_session_ident=7877148332545424738, latest_client_version=34, is_body_compressed=0, uncompressed_body_size=0, compressed_body_size=0)
2017-01-23 18:53:07.410284 RMPTest[2177:456331] Sync: Using already open Realm file: /var/mobile/Containers/Data/Application/02C61C4E-8A73-4EB7-83C2-5205C0AE6E7D/Documents/realm-object-server/8e55789bac5b692cb576c72fa6b33f39/realm%3A%2F%2F159.203.106.195%3A9080%2F%7E%2Frealmtasks
2017-01-23 18:53:07.417846 RMPTest[2177:456331] Sync: Connection[1]: WebSocket::frame_reader_loop()
2017-01-23 18:53:07.418115 RMPTest[2177:456331] Sync: Connection[1]: stage_delivery
2017-01-23 18:53:07.418895 RMPTest[2177:456331] Sync: Connection[1]: WebSocket: frame_reader.read_size = 2


If I run the app offline I only get the following log: 

> 2017-01-23 18:50:37.039213 RMPTest[2172:455051] [MC] System group container for systemgroup.com.apple.configurationprofiles path is /private/var/containers/Shared/SystemGroup/systemgroup.com.apple.configurationprofiles
2017-01-23 18:50:37.050385 RMPTest[2172:455051] [MC] Reading from public effective user settings.

**Comment by srpadronm** [comment]
@stel
Please also note that if I start the app with internet connection, then go offline, add some records and  reconnect internet again, everything get sync without any problems. 
The sync problem only appears if I start the app offline.

**Comment by stel** [comment]
Thanks for the info @srpadronm! It's probably a bug, we'll take a closer look on it.

**Comment by Montrazul** [comment]
Yes. I have to verify the fact that the app is never syncing if you login offline and the network connection comes back online. You have to close and reopen the app with existing internet connection to make the changes sync to the server. That could be a problem.

**Comment by srpadronm** [comment]
I login into the app with internet connection (at this point the app sync with the server perfectly), then if I go offline make some changes and go online again the app doesn’t sync with the server again.   I am supposed close and reopen the app every time I lost internet connection?

**Comment by Montrazul** [comment]
I guess this is a bug. I mean... if the rare case kicks in that a user after he changed data in offline mode, goes online and deletes the app after it... he would lose the data. So, lets wait for an fix.

**Comment by srpadronm** [comment]
Thanks

**Comment by Dids** [comment]
Is this still an issue?

**Comment by mkran** [comment]
I would like my users to be able to login offline to continue using my iOS app (assuming that they've signed up already). I'm using Realm DB as the mobile database and have set up sync with the Realm Object Server running in the Cloud. Is this possible today out of the box?

**Comment by jsflax** [comment]
"Offline login" is not something on our roadmap. However, any user that has logged in whilst online can still be accessed if they go offline.
